ipd/0020/README.adoc

```diff
@@ -0,0 +1,889 @@
+:tabsize: 8
+:toc:
+:toclevels: 5
+
+= IPD 20 Kernel Test Facility
+
+|===
+|Authors |State
+
+|Ryan Zezeski <ryan@zinascii.com>
+|draft
+|===
+
+
+== Goal
+
+Add a general testing facility to the kernel to allow for in situ
+kernel testing.
+
+== Background
+
+Testing is a vital part of any well maintained program, and a kernel,
+along with its various modules, should be no exception. But testing
+kernel modules can be hard, especially if trying to exercise specific
+branches in specific functions. For the most part we rely on clever
+userspace tests, in the form of `usr/src/test`, to exercise our most
+important code paths. While this can be made to work it also requires
+additional effort (as compared to what I'm about to propose) from the
+developer, if it can be made to work at all. Furthermore, these tests
+may be at the mercy of several layers of abstraction on top of the
+actual system they are trying to test. In short, it can sometimes feel
+like trying to drain a five gallon bucket with a straw.
+
+For many of us coming to kernel development for the first time there
+is often a moment of clarity when we realize the kernel is just
+another program. Sure, it's a "special" program, responsible for
+controlling the hardware and bringing order to the city of userspace
+programs, but at the end of the day its a program nonetheless. It's
```

Comment:
I think the apostrophe is missing here:
"its a program" => "it's a program"

Reply:
Good catch, turns out there were a few other its/it's mixups. Think I got them all this time.